Android - Core Integration

GhenadiePusca opened the pull request:
Somewhat big-ish PR, but it properly encapsulates the big picture.

What was done:

- Created JNI API for the core. C interface cannot be used since JNI requires a particular API format.
- Moved Rust FlatBuffers mapping in separate module so it can be shared between c_interface and android_interface modules.
- Created `PolyPodCoreAndroid` library meant to encapsulate the `core` interaction through JNI:

    - It encapsulates the FlatBuffers source files.
    - It encapsulates the generated FlatBuffer models.
    - It encapsulates the `jni` build libraries.
    - It describes the `jni` api, used to interact with native core.
    - It encapsulates FlatBuffer mapping logic.
- Created the `build_for_android.sh` script to be used to build the `core` for android usage:
    
     - It downloads the proper NDK version.
     - It links the necessary FlatBuffer source files with `PolyPodCoreAndroid`.
     - It links the generated FlatBuffer models `PolyPodCoreAndroid`.
     - It builds the core for the necessary architectures.
     - It links the built libraries `PolyPodCoreAndroid`.
- Linked `PolyPodCoreAndroid` with `PolyPod` app:

    - Bootstrap core at app startup.
    - Parse feature manifest by using `PolyPodCoreAndroid`
- Added new target in `core` Makefile used to build the android libraries. To not get the Makefile over-stuffed, I have decided to pull out the logic in `build_for_android.sh`, leaving the Makefile to act as a configuration/integration tool. The code for building iOS framework will be moved in its own script sometime later.
- Added the android core build step in relevant GH workflows.

GhenadiePusca commented:
Encapsulated the android core module in an Android project, so core be built and tested separately. At the same time, the PolyPodAndroid will just import the core module, not the whole project.

JJ commented on platform/android/app/src/main/kotlin/coop/polypoly/polypod/features/FeatureStorage.kt:
```diff
@@ -46,27 +47,22 @@ class FeatureStorage {
-            Language.determine(context)
-        )
+
+        return Core.parseFeatureManifest(manifestString).getOrNull()
```
> This just a placeholder for the time being

What happens to language? Before, context was needed to determine the language. Is that done now in the shared code?

fhd replied:
Yeah, the language is being passed to core upon initialisation (called "bootstrapping")

JJ commented on platform/core/PolyPodCoreAndroid/core/src/main/java/coop/polypoly/core/Core.kt:
```diff
@@ -0,0 +1,66 @@
+            when (response.resultType) {
+                FeatureManifestParsingResult.FeatureManifest -> {
+                    val manifest = response.result(FeatureManifest())
+                    if (manifest == null) {
```
This can be null even if parsing is correct?

fhd replied:
Yup, see below, this is just how Flatbuffers does things it seems.

GhenadiePusca replied:
Parsing to FlatBuffer model can fail as per FlatBuffer spec. So it will be failure of FlatBuffer parsing, not FeatureManifest parsing.

JJ commented on platform/core/PolyPodCoreAndroid/core/src/main/java/coop/polypoly/core/Core.kt:
```diff
@@ -0,0 +1,66 @@
+                            )
+                        )
+                    }
+                    return Result.success(manifest as FeatureManifest)
```
`manifest` is not already a `FeatureManifest`? Ah, OK, it can be `null`, I guess. But can it?

fhd replied:
It would appears so, `FeatureManifestParsingResponse` is part of the code generated by Flatbuffers, and the result can be `null` indeed.

GhenadiePusca replied:
`manifest` will be of type `Table` which has to be casted to `FeatureManifest`, it is not related to manifest being `null`.

JJ commented on platform/core/PolyPodCoreAndroid/core/src/main/java/coop/polypoly/core/Core.kt:
```diff
@@ -0,0 +1,66 @@
+                }
+                FeatureManifestParsingResult.Failure -> {
+                    val failure = response.result(Failure())
+                    if (failure == null) {
```
Failures can also fail?

GhenadiePusca replied:
`response.result` can fail, for example if Rust core did send bytes for a different type, other than `Failure`.

JJ commented on platform/core/PolyPodCoreAndroid/core/src/main/java/coop/polypoly/core/CoreExceptions.kt:
```diff
@@ -0,0 +1,30 @@
+
+import Failure
+
+@ExperimentalUnsignedTypes
```
`failure.code` may be unsigned? I don't see any other possible instance here.

fhd replied:
Funny enough, `failure.code` is `Byte`,  not `UByte`, so that doesn't seem to be it. But I guess we're accessing _some_ unsigned type below, otherwise this wouldn't be needed (and we'd get a warning I think).

GhenadiePusca replied:
Here `Failure` is used which is annotated with `ExperimentalUnsignedTypes`. The FlatBuffers model body can contain internal unsigned types. `Failure` itself does not have any, but Flatbuffers generator automatically adds the annotation.

JJ commented on platform/core/PolyPodCoreAndroid/core/src/main/java/coop/polypoly/core/CoreExceptions.kt:
```diff
@@ -0,0 +1,30 @@
+}
+
+class InvalidFailureContentException(context: String) :
+    Exception("$context -> received failure result type without content")
```
"without" or "with invalid"?

fhd replied:
It appears "without" is accurate - this is being used above when the whole `failure` object is `null`.

JJ replied:
The thing is that the class is actually called "Invalid". So I guess it's more of a naming nit.

GhenadiePusca replied:
True,`MissingFailureContentException` would be a better name.

JJ commented on platform/core/PolyPodCoreAndroid/core/src/main/java/coop/polypoly/core/CoreExceptions.kt:
```diff
@@ -0,0 +1,30 @@
+    Exception("$context -> received failure result type without content")
+
+class InvalidFeatureManifestContentException(context: String) :
+    Exception("$context -> received feature manifest result type without content")
```
Same as above, "without" or "with invalid"?

JJ commented on platform/core/PolyPodCoreAndroid/core/src/test/java/coop/polypoly/core/ExampleUnitTest.kt:
```diff
@@ -0,0 +1,16 @@
+import org.junit.Test
+
+/**
+ * Example local unit test, which will execute on the development machine (host).
```
nit: "run on"

JJ commented on platform/core/PolyPodCoreAndroid/core/src/test/java/coop/polypoly/core/ExampleUnitTest.kt:
```diff
@@ -0,0 +1,16 @@
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+class ExampleUnitTest {
```
Sorry, what's the point of this?

fhd replied:
This is just an example file generated by Idea it appears - let's just remove it for now I'd say. Adding some smoke tests for the bindings would definitely be useful though.

GhenadiePusca replied:
Yeah, it could be removed. As for the tests for bindings, I still have to think to which extent we want to test the interaction with `core` on different platforms. I would not want to have duplicated tests between platforms.

JJ commented on platform/core/PolyPodCoreAndroid/gradle.properties:
```diff
@@ -0,0 +1,23 @@
+# http://www.gradle.org/docs/current/userguide/build_environment.html
+# Specifies the JVM arguments used for the daemon process.
+# The setting is particularly useful for tweaking memory settings.
+org.gradle.jvmargs=-Xmx2048m -Dfile.encoding=UTF-8
```
Couldn't you add some "UnsignedExperimentalWhatever" here to not need to insert it in specific files?
Also, what's the best practice regarding these experimental features?

fhd replied:
These are the JVM parameters, i.e. runtime. The `ExperimentalUnsignedTypes` parameter is an option for the Kotlin compiler. So no :D

JJ commented on platform/core/build_for_android.sh:
```diff
@@ -0,0 +1,101 @@
+	esac
+
+	ndk_download_link=$google_repo/$ndk_lib-$os_name-x86_64.zip 
+	curl -L $ndk_download_link -o ${ndk_dir}.zip
```
You can call this file anything you want, as long as you use it throughout. Calling it `ndk_dir` is a bit confusing to me. But not a big deal.

GhenadiePusca replied:
Yeah, now after the vacation, I see what you mean :)

JJ reviewed (commented):
Since this is still in draft, I can't go for "Approve", but ti don't see any issue with what's been done. Great job.

fhd commented on .github/workflows/ios-android-test.yml:
```diff
@@ -33,7 +33,7 @@ jobs:
             - name: Build API and features
               run: ./build.js
-            - name: Build core
+            - name: Build iOS core
```
I've changed the build process for core in `main` - it now builds the platform specific bindings automatically (iOS only when the build happens on macOS), so _Build core_ will be accurate :) We can rather get rid of _Build Android core_ below. I'll take this branch over to merge in `main` and finish the job.

fhd commented on platform/android/app/src/main/kotlin/coop/polypoly/polypod/MainActivity.kt:
```diff
@@ -4,16 +4,32 @@ import android.app.AlertDialog
+    companion object {
+        @Suppress("JAVA_CLASS_ON_COMPANION")
+        private val logger = LoggerFactory.getLogger(javaClass.enclosingClass)
+    }
```
It's how logging works all over the code base already - we can probably move to a Kotlin-only logging framework, but thoroughly out of scope here I'd argue :)

fhd commented:
Since Ghenadie is out a little while longer, I'll work on this a little more - at least resolving the conflicts (caused by me :P), but ideally getting it ready to be merged.

fhd commented on platform/core/build_for_android.sh:
```diff
@@ -0,0 +1,101 @@
+	    ;;
+	esac
+
+	ndk_download_link=$google_repo/$ndk_lib-$os_name-x86_64.zip 
```
Would that actually work on M1 macbooks?

I'm pretty sceptical about fetching the NDK as part of the build - it's like 1GB... But I think I wanna leave it like that at first.

JJ replied:
Don't M1 machines emulate that architecture? 
Would be nice to have a native version, anyway...

GhenadiePusca replied:
As far as I know, Apple M1 support is available starting with NDK version r23. But with rust we still have to use NDK r22. There are some workaround to make Rust properly work with NDK `>=r23`, but I think it is better to use stable `r22` version. Later versions of NDK, `>=r23` do not require specifying the architecture, so when NDK version will be updated, we will remove the architecture parameter.

Yes, it works on M1 MacBooks through Rosetta, and it is the accepted way to run NDK `<r23` version.

fhd commented:
I'm not going to get around to do a lot of work on this, but I want to at least:

1. Merge in `main`
2. Streamline the Makefile a bit - similar to the logic we have now for iOS, I would build the Android bindings automatically, to reduce the amount of manual work people have to do to build everything - so just `make -C core` does the trick.
3. Make it easier to build only _ios_, only _android_, or even only _core_ specifically, with dedicated phony targets (which would depend on actual files, the way I already did it for iOS).

From a quick look, that's the main stuff I'd want to change, but I'm also gonna at least do a review and leave some comments before I fix other stuff I think.

fhd commented:
OK, now the build logic is a little less convoluted - but still pretty convoluted :D Having all the Android-specific build logic in `build_for_android.sh` really isn't helping. In general, it's a bit tricky to mix makefiles and scripts like this, makes it pretty difficult to provide incremental builds. But I think for the purposes of this PR, it's almost good enough.

One change I would still consider useful - the way the NDK is installed:

1. Let the user set an `$ANDROID_NDK_DIR` environment variable.
2. Only fetch the NDK automatically (to that location) if it doesn't exist, otherwise skip the fetching step.
3. If the `$ANDROID_NDK_DIR` variable is empty, abort with an error.

This way we'd achieve several things:

1. The user can get the NDK they actually need (currently x86_64 is hard coded, we'd need to do architecture detection to find the right one, I think it's fair to ask the user to manually install it and get rid of the automatic fetching I think...).
2. The NDK is somewhat large even by modern standards - 1 GB compressed, don't think we should surprise users with such a download.
3. The NDK wouldn't be subject to `make clean`, since it'd be outside the repository - where it arguably belongs.

fhd commented on platform/android/app/build.gradle:
```diff
@@ -107,6 +107,7 @@ repositories {
 
 dependencies {
     implementation project(":polyOut")
+    implementation project(":PolyPodCoreAndroid:core")
```
Same thought - for the _polyOut_ module, it appears to work without having an extra module it seems, would be nice if we had the same setup here.

GhenadiePusca replied:
`core` module was wrapped in PolyPodCoreAndroid so that we are able to build and test it separately from PolyPodAndroid, the same way we can do with PolyPodCoreSwift; so one does not have to build PolyPodAndroid in order to build and test `core`.

I think I can name the project just `android`, since the actual module that is integrated in PolyPodApp is `core`.

GhenadiePusca replied:
Pointed directly to `core` module instead of going through `PolyPodCoreAndroid`.

fhd commented on platform/android/app/src/main/kotlin/coop/polypoly/polypod/MainActivity.kt:
```diff
@@ -4,16 +4,32 @@ import android.app.AlertDialog
+        val language = Language.determine(this@MainActivity)
+        val bootstrapResult = Core.bootstrapCore(language)
+        if (bootstrapResult.isSuccess) {
+            logger.info("Core is bootstraped!")
```
Yup! For strings that can actually show up in the app, we have `dev_strings.xml`, but for things that only show up in the log, we just hard code English so far, it's probably fine.

fhd commented on platform/android/app/src/main/kotlin/coop/polypoly/polypod/MainActivity.kt:
```diff
@@ -4,16 +4,32 @@ import android.app.AlertDialog
+            logger.error(
+                "Failed to boostrap core",
+                bootstrapResult.exceptionOrNull()
+            )
```
I feel like we're asking for trouble here - you don't actually quit the app, so we're just going to have undefined behaviour at runtime? Seems like in this case the app would crash when trying to show the feature list. That's gonna be nasty to debug...

How about we skip any logging and just go for:

```
if (!bootstrapResult.isSuccess) {
    throw new RuntimeException("Failed to bootstrap core: ${bootstrapResult.exceptionOrNull}")
}
```

? That way, we deliberately crash the app with a specific message, which should make it a lot easier to identify in crash reports.

GhenadiePusca replied:
True, it will be safer to crash the app.

fhd commented on platform/android/app/src/main/kotlin/coop/polypoly/polypod/features/Feature.kt:
```diff
@@ -9,24 +10,24 @@ import java.util.zip.ZipFile
     val fileName: String,
     val content: ZipFile,
-    private val manifest: FeatureManifest
+    private val manifest: FeatureManifest?
```
Well, we do support features without manifest, but in the past we still had a `FeatureManifest` object - now it appears there's situations where we don't. But if you look at the code below, there was always some logic for figuring out the most important stuff (like the feature name) with no manifest.

I think that this default logic should _probably_ go into the manifest parsing logic, so we don't duplicate it for each platform, but we can handle that in a followup.

GhenadiePusca replied:
This is based on the current specs, I merely converted a FeatureManifest full of `null` values to optional FeatureManifest.

fhd commented on platform/android/app/src/main/kotlin/coop/polypoly/polypod/features/Feature.kt:
```diff
@@ -36,9 +37,16 @@ class Feature(
-        in manifest.links?.keys ?: listOf() -> manifest.links?.get(target)
-        in manifest.links?.values ?: listOf() -> target
-        else -> null
+    fun findUrl(target: String): String? {
```
I'd say this link lookup logic should also move into platform core, but we can do that in a followup.

GhenadiePusca replied:
I did update it to match the iOS code:
- If `target` matches the link key then return the translated url.
- If `target` matches translated url of any link, then return the target.
- If no match, return nil.

And yeah, this code could be in core, so it is shared and consistent between platforms. But this will be a followup.

fhd commented on platform/android/app/src/main/kotlin/coop/polypoly/polypod/features/FeatureStorage.kt:
```diff
@@ -46,27 +47,22 @@ class FeatureStorage {
-            return FeatureManifest(
-                null, null, null, null, null, null, null, null
-            )
+            return null
```
@JJ This is presumably why manifest was made optional above - previously we would return, logically, an empty manifest - now it's just `null`, presumably because we can't create an empty manifest so easily anymore.

fhd commented on platform/core/PolyPodCoreAndroid/.idea/.gitignore:
```diff
@@ -0,0 +1,3 @@
+# Default ignored files
```
We should really find a way to ignore the entire `.idea` directory - this is first and foremost a Gradle project, Idea/Android Studio is just one possible IDE.

fhd commented on platform/core/PolyPodCoreAndroid/core/src/main/java/coop/polypoly/core/Core.kt:
```diff
@@ -0,0 +1,66 @@
+@ExperimentalUnsignedTypes
+class Core {
+    companion object {
+        fun bootstrapCore(languageCode: String): Result<Unit> {
```
For the record, I'm not a huge fan of having to deal with two different error handling mechanisms - in the Android app we use exceptions consistently, now in the core Android bindings we're throwing around result objects. I think it's OK-ish because it's a separate module at the end of the day, but felt like pointing it out anyway :)

GhenadiePusca replied:
Agree, I missed that detail - that PolyPodAndroid use Exceptions instead. Better we keep it consistent, I will update the code :)

fhd commented on platform/android/app/build.gradle:
```diff
@@ -257,3 +258,10 @@ tasks.whenTaskAdded { task ->
+tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).configureEach {
+    kotlinOptions {
+        // Generated FlatBuffer models use ExperimentalUnsignedTypes
+        freeCompilerArgs += "-opt-in=kotlin.ExperimentalUnsignedTypes"
```
I find it non-trivial to check what Kotlin version we're on, but it appears that if we are on >=1.5, and potentially also a newer version of Flatbuffers, we won't need to do this, see: https://github.com/google/flatbuffers/issues/7087

GhenadiePusca replied:
That's correct, but latest FlattBuffers release with precompiled `flatc` is `2.0.0`. And we want to avoid compiling `flatc` locally :)

fhd reviewed (changes requested):
Alright, so I know it's still in _Draft_, but I think it's actually reasonably close to being ready to land, nice work!

Here's the major stuff I think we should address:

1. Split out the NDK download - and test it on an M1 MacBook (since we used to hard code x86)
2. Eliminate the `.idea` directory of the PolyPodAndroidCore project from the repo
3. Remove that example test in the PolyPodAndroidCore project
4. See if we can't get rid of the extra _core_ module (or whatever it is) in PolyPodAndroidCore
5. Deliberately crash app when core initialisation fails

There's also inline comments for all of these things, but I'm too lazy to link to them now :D

fhd commented:
For the remaining shell script / makefile mess, I created a followup ticket: https://jira.polypoly.eu/browse/PROD4POD-1548 -> This PR is leaving the messiness at around the level it was before, so I find it out of scope to fix any of that here.

GhenadiePusca commented on .github/workflows/js-and-android-test.yml:
```diff
@@ -62,6 +62,9 @@ jobs:
                   cache: 'gradle'
+            - name: Build Android core
+              working-directory: platform/core
+              run: make android_libraries
```
Yeah it could be an improvement, I am not yet very familiar with how caching artifacts between workflow works in GH actions.

GhenadiePusca commented on .github/workflows/js-and-android-test.yml:
```diff
@@ -62,6 +62,9 @@ jobs:
                   java-version: 11
                   distribution: 'temurin'
                   cache: 'gradle'
+            - name: Build Android core
```
@fhd I am not sure I understand how this is gone. I don't see any changes that caches the artefacts between workflows.

GhenadiePusca commented:
> OK, now the build logic is a little less convoluted - but still pretty convoluted :D Having all the Android-specific build logic in `build_for_android.sh` really isn't helping. In general, it's a bit tricky to mix makefiles and scripts like this, makes it pretty difficult to provide incremental builds. But I think for the purposes of this PR, it's almost good enough.
> 
> One change I would still consider useful - the way the NDK is installed:
> 
> 1. Let the user set an `$ANDROID_NDK_DIR` environment variable.
> 2. Only fetch the NDK automatically (to that location) if it doesn't exist, otherwise skip the fetching step.
> 3. If the `$ANDROID_NDK_DIR` variable is empty, abort with an error.
> 
> This way we'd achieve several things:
> 
> 1. The user can get the NDK they actually need (currently x86_64 is hard coded, we'd need to do architecture detection to find the right one, I think it's fair to ask the user to manually install it and get rid of the automatic fetching I think...).
> 2. The NDK is somewhat large even by modern standards - 1 GB compressed, don't think we should surprise users with such a download.
> 3. The NDK wouldn't be subject to `make clean`, since it'd be outside the repository - where it arguably belongs.

OK, so I might have rushed this PR a little bit, in order to have it for intermediary review before going to vacation, and I might have missed to explain some of the choices that were made.

- The reason I have moved the android building logic in a separate shell script is to cleanup the Makefile. My idea is to have the heavy lifting in the shell script(which is more portable than Makefile), and use the Makefile for the means of specifying dependencies and configuration. Based on the current build logic, what can be possible factored out is the NDK installation, which can be a separate step.
- Rust core needs to use a specific version of NDK, not just any. It is downloaded and stored locally in the repo so it does not conflict with whatever NDK version is installed globally on the machine. I think it would be trickier for someone to manually download a separate version of NDK, which is not conflicting with whatever other version they have, in order to build core.
- As for architecture difference, note that according to https://github.com/android/ndk/wiki/Unsupported-Downloads, the only supported architecture for NDK `<=r22` is `x86_64`, there are no other options. If one would want to support other architectures(which is highly unlikely), it might be needed build the NDK from scratch for particular architecture.

GhenadiePusca commented:
> Alright, so I know it's still in _Draft_, but I think it's actually reasonably close to being ready to land, nice work!
> 
> Here's the major stuff I think we should address:
> 
> 1. Split out the NDK download - and test it on an M1 MacBook (since we used to hard code x86)
> 2. Eliminate the `.idea` directory of the PolyPodAndroidCore project from the repo
> 3. Remove that example test in the PolyPodAndroidCore project
> 4. See if we can't get rid of the extra _core_ module (or whatever it is) in PolyPodAndroidCore
> 5. Deliberately crash app when core initialisation fails
> 
> There's also inline comments for all of these things, but I'm too lazy to link to them now :D

I have addressed all of your points besides # 1 - see [reason](https://github.com/polypoly-eu/polyPod/pull/820#discussion_r864648671) and # 4 - see [reason](https://github.com/polypoly-eu/polyPod/pull/820#discussion_r864608835).

I have made some other smaller adjustments, and I this looks good to land.

And of course, thank you @fhd for making some very nice improvements 💯 !

JJ commented:
It's now conflicting, can you take a look?

GhenadiePusca commented:
> It's now conflicting, can you take a look?

I am also waiting for this https://github.com/polypoly-eu/polyPod/pull/869 to land.

JJ commented on platform/core/PolyPodCoreAndroid/core/build.gradle:
```diff
@@ -0,0 +1,48 @@
+
+android {
+    compileSdk 30
+    buildToolsVersion "30.0.3"
```
This [SO answer](https://stackoverflow.com/a/51960054/891440) seems to recommend some `ext` extra properties declared at the top level; that would be `platform`, I guess; that would probably imply carrying over the gradle scripts there; it, however, makes sense to have this kind of thing, since with this we will need integrated build scripts at the platform level. Possibly outside the scope of what is already a huge PR.

JJ commented on platform/core/PolyPodCoreAndroid/core/proguard-rules.pro:
```diff
@@ -0,0 +1,21 @@
+# Add project specific ProGuard rules here.
```
What is this?

JJ commented on platform/core/README.md:
```diff
@@ -28,6 +33,10 @@ Swift bindings), as follows:
 
+- Build the Android bindings:
+
+        ./build_for_android.sh
```
I don't know if make can't be used. In the other PR we already assume people will have it installed somehow, even in Windows.

GhenadiePusca replied:
I have refactored the makefile to also include android build, thus removing the shell script.

JJ reviewed (approved):
Good job for a major change. :100:

JJ commented:
`android_bindings` not found...

GhenadiePusca commented:
> `android_bindings` not found...

I am currently working on it

GhenadiePusca commented on platform/android/app/src/main/kotlin/coop/polypoly/polypod/features/Feature.kt:
```diff
@@ -9,24 +10,24 @@ import java.util.zip.ZipFile
-    val name: String get() = manifest.name ?: id
-    val author: String get() = manifest.author ?: ""
-    val description: String get() = manifest.description ?: ""
+    val name: String get() = manifest?.name ?: id
```
Based on in the implementation, the feature manifest can be null if Parser failed; idk why, but for some reason we want to display the feature even if its manifest was not parsed. Normally those should be filtered out, but this is how it was implemented previously. We can discuss about this in a Jira ticket, I'll create one.

GhenadiePusca commented on .github/workflows/upload_all.yml:
```diff
@@ -123,7 +126,7 @@ jobs:
       - name: iOS - Build core
         working-directory: platform/core
-        run: make MODE=release
+        run: make ios_bindings MODE=release
```
In `ios-android-test.yml` we just execute `make`, which would build both android and iOS bindings.

GhenadiePusca commented on platform/android/README.md:
```diff
@@ -4,8 +4,13 @@ This directory contains the source for the polyPod app for Android devices.
-the [`README` file](../../README.md) above.
+### Prerequisites
+
+The main prerequisite you need are the polyPod features and the feature API -
```
I will update to `... polyPod feature built ...`

JJ commented on platform/core/README.md:
```diff
@@ -9,10 +9,14 @@ Logic shared across polyPod platforms - ios, android, wasm.
+If you wish to only build the core binaries for your current operating system,
+run `make core`. If you only wish to build the Android binaries and bindings,
+run `make android_bindings`. Similarly, to only build the iOS binaries and
+bindings, run `make ios_bindings`.
```
Maybe clarify that you can only run both in MacOS, since ios_bindings can't work in Linux or Windows, if I'm not wrong.

GhenadiePusca replied:
Sure I can, at the same time `make` will return early with an error telling that only macOS can be used.

GhenadiePusca replied:
Sure I can, at the same time `make` will return early with an error telling that only macOS can be used.

JJ commented on platform/core/Makefile:
```diff
@@ -181,8 +181,8 @@ endef
 
-.PHONY: rust-core
-rust-core: $(target)
+.PHONY: rust_core
```
You're normalizing to snake_case all targets, right?

GhenadiePusca replied:
Yes.

JJ replied:
Good call.

JJ commented on platform/core/PolyPodCoreAndroid/core/build.gradle:
```diff
@@ -0,0 +1,46 @@
+        reporter "plain"
+    }
+    filter {
+        exclude("**/FlatbuffersGenerated/**")
```
Let me see if I understand what you're doing here. Essentially, you're excluding a specific path from ktlint, which is done at the top level. So is this really used? Wouldn't you need to do this at the top level configuration?

GhenadiePusca replied:
I don't quite understand what you mean by `top level configuration`?

JJ replied:
The `ktlint` configuration we (might) have at the root or `platform/android` directory. We don't actually have such thing, mainly because we don't need it... But if we did, my point is that possibly that should be included in the root directory somehow so that it's actually picked up by ktlint and its workflow.

GhenadiePusca replied:
I understand, make sense.
As for `is it used` - I missed to run `ktlint` workflow also for android core. I'll add it.

GhenadiePusca replied:
Just tested https://github.com/polypoly-eu/polyPod/runs/6466721501?check_suite_focus=true and ktlint was executed on the android core.

JJ reviewed (approved):
Still good. Great work.

JJ commented on platform/android/app/src/test/kotlin/coop/polypoly/polypod/polyout/PolyOutTest.kt:
```diff
@@ -9,7 +9,6 @@ import coop.polypoly.polypod.features.FeatureStorage
 import coop.polypoly.polypod.polyOut.PolyOut
 import coop.polypoly.polypod.util.FakeAesKeyGenerator
 import coop.polypoly.polypod.util.FakeKeyStore
-import coop.polypoly.polypod.util.MockFeature
```
But this means that it's not used anywhere, right? So the code can be removed too.
